src/test/java/org/apache/commons/lang3/concurrent/AbstractConcurrentInitializerCloseAndExceptionsTest.java

```diff
@@ -0,0 +1,189 @@
+     * ConcurrentException it will rethrow it wrapped in a ConcurrentException
+     */
+    @SuppressWarnings("rawtypes")
+    @Test
+    public void testCloserThrowsCheckedException() throws ConcurrentException {
+        final ConcurrentInitializer<CloseableObject> initializer = createInitializerThatThrowsException(
+                CloseableObject::new,
+                (CloseableObject) -> methodThatThrowsException(ExceptionToThrow.IOException));
+
+        try {
+            initializer.get();
+            ((AbstractConcurrentInitializer) initializer).close();
+            fail();
+        } catch (Exception e) {
+            assertThat(e, instanceOf(ConcurrentException.class));
+            assertThat(e.getCause(), instanceOf(IOException.class));
+        }
+    }
+
+    /**
+     * This method tests that if AbstractConcurrentInitializer.close catches a
+     * RuntimeException it will throw it withuot wrapping it in a ConcurrentException
+     */
+    @SuppressWarnings("rawtypes")
+    @Test
+    public void testCloserThrowsRuntimeException() throws ConcurrentException {
+        final ConcurrentInitializer<CloseableObject> initializer = createInitializerThatThrowsException(
+                CloseableObject::new,
+                (CloseableObject) -> methodThatThrowsException(ExceptionToThrow.NullPointerException));
+
+        initializer.get();
+        assertThrows(NullPointerException.class, () -> {
+            ((AbstractConcurrentInitializer) initializer).close();
+            });
+    }
+
+    /**
+     * This method tests that if AbstractConcurrentInitializer.close actually closes the wrapped object
+     */
+    @SuppressWarnings("rawtypes")
+    @Test
+    public void testWorkingCloser() throws Exception {
+        final ConcurrentInitializer<CloseableObject> initializer = createInitializerThatThrowsException(
+                CloseableObject::new,
+                CloseableObject::close);
+
+        CloseableObject cloesableObject = initializer.get();
+        assertFalse(cloesableObject.isClosed());
+        ((AbstractConcurrentInitializer) initializer).close();
+        assertTrue(cloesableObject.isClosed());
+    }
+
+    protected enum ExceptionToThrow {
+        IOException,
+        SQLException,
+        NullPointerException
+    }
+
+    //The use of enums makes this look like a realistic method to the compiler
+    protected static CloseableObject methodThatThrowsException(ExceptionToThrow input) throws IOException, SQLException, ConcurrentException {
+        switch (input) {
+        case IOException:
+            throw new IOException();
+        case SQLException:
+            throw new SQLException();
+        case NullPointerException:
+            throw new NullPointerException();
+        default:
+            fail();
+            return new CloseableObject();
+        }
+    }
+
+    protected abstract ConcurrentInitializer<CloseableObject> createInitializerThatThrowsException(
```

Comment:
It looks like all implementations of `createInitializerThatThrowsException()` and `createInitializerThatThrowsPreCreatedException()` are the same for each subclass. Am I seeing this right? If so, do we need the flexibility for future tests or can this be simplified to one abstract method?

Reply:
That is amusing. They started of different and were made more flexible until they became identical so we only need the one.